src/ui/features/onboarding/Success/WebAppMessageHandler.tsx

```diff
@@ -0,0 +1,84 @@
+import React, { useEffect, useRef } from 'react';
+import { invariant } from 'src/shared/invariant';
+import { isObj } from 'src/shared/isObj';
+
+const ZERION_WEB_APP_URL = new URL(
+  process.env.NODE_ENV === 'development'
+    ? 'http://localhost:3000'
+    : 'https://app.zerion.io'
+);
+
+enum WebAppCallbackMethod {
+  SetReferralCode = 'set-referral-code',
+}
+
+interface WebAppMessage<T = unknown> {
+  method: WebAppCallbackMethod;
+  params?: T;
+}
+
+export function isWebAppMessage(
+  event: MessageEvent,
+  {
+    expectedSource,
+  }: {
+    expectedSource: Window | null;
+  }
+): event is MessageEvent<WebAppMessage> {
+  return (
+    event.origin === ZERION_WEB_APP_URL.origin &&
+    event.source === expectedSource &&
+    isObj(event.data) &&
+    'method' in event.data
+  );
+}
+
+function setReferralCode(_referralCode: string) {
+  // save referralCode
```

Comment:
Here we also need not to safe incorrect code to avoid softlock state for the user who clicked on incorrect link

Reply:
Now referral code is checked on the webapp side, so it should be safe to just save it here as is

Reply:
I think, doublecheck will not harm :)

Reply:
Extension should be more reliable then web